components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,256 @@
+                {
+                    set_panic_mark_and_panic(
+                        path,
+                        &format!(
+                            "key range mismatch, smallest key:{:?}, largest key:{:?}",
+                            &f.smallest_key, &f.largest_key
+                        ),
+                    );
+                }
+
+                // lock the store_meta and check if the range exist.
+                if !self.check_overlap_damaged_regions(f) {
+                    set_panic_mark_and_panic(path, "Can't find overlap regions");
+                }
+
+                return;
+            }
+        }
+    }
+
+    // `sst_path` has been processed and is still in a damaged state.
+    fn exist_scheduling_regions(&self, sst_path: &str) -> bool {
+        self.damaged_files.iter().any(|f| f.name == sst_path)
+    }
+
+    // Periodically check for recorded damaged files。
+    //
+    // Acquire meta lock.
+    fn check_damaged_files(&mut self) {
+        let mut new_damaged_files = self.damaged_files.clone();
+        new_damaged_files.retain(|f| {
+            if f.start_time.elapsed() > self.max_hang_duration {
+                set_panic_mark_and_panic(
+                    &f.name,
+                    &format!("recovery job exceeded {:?}", self.max_hang_duration),
+                );
+            }
+            let need_retain = self.check_overlap_damaged_regions(f.clone());
+            if !need_retain {
+                println!(
+                    "ready delete_file {} {:?} {:?} (this line must be deleted before merge PR)",
+                    f.name, &f.smallest_key, &f.largest_key
+                );
+                // set `include_end` to `true` to ensure that sst with the same start key and end key.
+                self.db
+                    .delete_files_in_range(&f.smallest_key, &f.largest_key, true)
+                    .unwrap();
+            }
+            need_retain
+        });
+        self.damaged_files = new_damaged_files;
+    }
+
+    // Check whether the StoreMeta contains the region range, if it contains,
+    // recorded fault region ids to report to PD and add file info into `damaged_files`.
+    //
+    // Acquire meta lock.
+    fn check_overlap_damaged_regions(&mut self, mut file: FileInfo) -> bool {
+        let mut meta = self.store_meta.lock().unwrap();
+
+        // Find overlapping region ids.
+        // Condition:
+        // end_key > file.smallestkey
+        // start_key <= file.largestkey
+        let mut ids = vec![];
+        for (_, id) in meta
+            .region_ranges
+            .range((Excluded(file.smallest_key.clone()), Unbounded::<Vec<u8>>))
+        {
+            let region = &meta.regions[id];
+            if keys::enc_start_key(region) <= file.largest_key {
+                ids.push(*id);
+            }
+        }
+
+        let overlap = !ids.is_empty();
+        if overlap && !self.exist_scheduling_regions(&file.name) {
```

Comment:
if the file is detected before, but one region within the range is split. Then you should check the ids again to push the newly created one to `damage_regions_id`.